lib/std/Build.zig

```diff
@@ -1559,16 +1657,79 @@ pub fn anonymousDependency(
     return dependencyInner(b, name, build_root, build_zig, args);
 }
 
+fn userValuesAreSame(lhs: UserValue, rhs: UserValue) bool {
+    switch (lhs) {
+        .flag => {},
+        .scalar => |lhs_scalar| {
+            const rhs_scalar = switch (rhs) {
+                .scalar => |scalar| scalar,
+                else => return false,
+            };
+
+            if (!std.mem.eql(u8, lhs_scalar, rhs_scalar))
+                return false;
+        },
+        .list => |lhs_list| {
+            const rhs_list = switch (rhs) {
+                .list => |list| list,
+                else => return false,
+            };
+
+            if (lhs_list.items.len != rhs_list.items.len)
+                return false;
+
+            for (lhs_list.items, rhs_list.items) |lhs_list_entry, rhs_list_entry| {
+                if (!std.mem.eql(u8, lhs_list_entry, rhs_list_entry))
+                    return false;
+            }
+        },
+        .map => |lhs_map| {
+            const rhs_map = switch (rhs) {
+                .map => |map| map,
+                else => return false,
+            };
+
+            if (lhs_map.count() != rhs_map.count())
+                return false;
+
+            var lhs_it = lhs_map.iterator();
+            while (lhs_it.next()) |lhs_entry| {
+                const rhs_value = rhs_map.get(lhs_entry.key_ptr.*) orelse return false;
+                if (!userValuesAreSame(lhs_entry.value_ptr.*.*, rhs_value.*))
+                    return false;
+            }
+        },
+    }
+
+    return true;
+}
+
+fn userInputOptionsMapsAreSame(lhs: UserInputOptionsMap, rhs: UserInputOptionsMap) bool {
+    if (lhs.count() != rhs.count())
+        return false;
+
+    var it = lhs.iterator();
+    while (it.next()) |lhs_entry| {
+        const rhs_value = rhs.get(lhs_entry.key_ptr.*) orelse return false;
+        if (!userValuesAreSame(lhs_entry.value_ptr.*.value, rhs_value.value))
+            return false;
+    }
+
+    return true;
+}
+
 pub fn dependencyInner(
     b: *Build,
     name: []const u8,
     build_root_string: []const u8,
     comptime build_zig: type,
     args: anytype,
 ) *Dependency {
+    const user_input_options = userInputOptionsFromArgs(b.allocator, args);
     if (b.initialized_deps.get(build_root_string)) |dep| {
-        // TODO: check args are the same
-        return dep;
+        if (userInputOptionsMapsAreSame(user_input_options, dep.builder.user_input_options)) {
+            return dep;
+        }
```

Comment:
hmm I think I see a big problem here. This is correctly checking if a value in a hash map can be re-used, but then at the end of the function it will put the new value into the hash map, overwriting the old one. In other words, it does not successfully deduplicate with more complicated set of inputs. If we passed ABA, for example, it would create two A's instead of one A and one B.

I think you already created all the glue code that is needed to solve this; the `initialized_deps` hash map needs to key on not only the build_root_string but on the user input options as well. That hash map is exclusively used in this function, so you are free to change it to this function's needs.

Reply:
Thanks for pointing that out, I completely missed it. I’ve added user input to the key of `initialized_deps`.